sdks/python/apache_beam/runners/common.py

```diff
@@ -969,7 +1212,10 @@ def _reraise_augmented(self, exn):
 
 class OutputProcessor(object):
 
-  def process_outputs(self, windowed_input_element, results):
+  def process_outputs(self,
+                      windowed_input_element,
+                      results,
+                      watermark_estimator=None):
```

Comment:
Is there value to making this an optional argument or should we make it required?

Reply:
Or, even better, perhaps pass it in as a constructor argument (if it will never change)?

Reply:
> Is there value to making this an optional argument or should we make it required?
Without providing no-op estimator, for those DoFn without watermark_estimator provided, this argument is optional. If we decided to have no-op estimator, this argument then should be required.

> Or, even better, perhaps pass it in as a constructor argument (if it will never change)?
The estimator should stay the same for the element. If the OutputProcessor is created per element, then yes we can make it as constructor argument.